src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -115,25 +117,61 @@ public static ShardPath loadShardPath(ESLogger logger, NodeEnvironment env, Shar
+        long estShardSizeInBytes = Math.max(avgShardSizeInBytes, (long) (totFreeSpace/20.0));
+
+        // Collate predicted (guessed!) disk usage on each path.data:
+        Map<Path,Long> reservedBytes = new HashMap<>();
+        for (IndexShard shard : shards) {
+            Path dataPath = NodeEnvironment.shardStatePathToDataPath(shard.shardPath().getShardStatePath());
+
+            // Remove indices/<index>/<shardID> subdirs from the statePath to get back to the path.data/<lockID>:
+            Long curBytes = reservedBytes.get(dataPath);
+            if (curBytes == null) {
+                curBytes = 0L;
+            }
+            reservedBytes.put(dataPath, curBytes + estShardSizeInBytes);
+        }       
+
+        return reservedBytes;
+    }
+
+    public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shardId, @IndexSettings Settings indexSettings,
+                                                  long avgShardSizeInBytes, Iterable<IndexShard> shards) throws IOException {
+
+        Map<Path,Long> estReservedBytes = getEstimatedReservedBytes(env, avgShardSizeInBytes, shards);
+
+        // TODO - do we need something more extensible? Yet, this does the job for now...
         final String indexUUID = indexSettings.get(IndexMetaData.SETTING_UUID, IndexMetaData.INDEX_UUID_NA_VALUE);
         final NodeEnvironment.NodePath[] paths = env.nodePaths();
-        final List<Tuple<Path, Long>> minUsedPaths = new ArrayList<>();
+        NodeEnvironment.NodePath bestPath = null;
+        long maxUsableBytes = Long.MIN_VALUE;
         for (NodeEnvironment.NodePath nodePath : paths) {
-            final Path shardPath = nodePath.resolve(shardId);
             FileStore fileStore = nodePath.fileStore;
-            long usableSpace = fileStore.getUsableSpace();
-            if (minUsedPaths.isEmpty() || minUsedPaths.get(0).v2() == usableSpace) {
-                minUsedPaths.add(new Tuple<>(shardPath, usableSpace));
-            } else if (minUsedPaths.get(0).v2() < usableSpace) {
-                minUsedPaths.clear();
-                minUsedPaths.add(new Tuple<>(shardPath, usableSpace));
+            long usableBytes = fileStore.getUsableSpace();
+            Long reservedBytes = estReservedBytes.get(nodePath.path);
+            if (reservedBytes != null) {
+                // Deduct estimated reserved bytes from usable space:
+                usableBytes -= reservedBytes;
+            }
+            if (usableBytes > maxUsableBytes) {
+                maxUsableBytes = usableBytes;
+                bestPath = nodePath;
             }
         }
-        Path minUsed = minUsedPaths.get(shardId.id() % minUsedPaths.size()).v1();
+
+        final Path statePath = bestPath.resolve(shardId);
         final Path dataPath;
-        final Path statePath = minUsed;
+        
         if (NodeEnvironment.hasCustomDataPath(indexSettings)) {
```

Comment:
> if we have custom data path it can only be one path so there is nothing to select?

This is what I expected too, but go look at ShardPath.java right now in master: it still passes the loadedPath to the new ShardPath (as shardStatePath) even in the custom data path case.  This confuses me :)

I think it means we store shard state (ShardStateMetaData) on the local node's path.data even if custom data path is set?